Add a static cache that offloads to the CPU or other device

gerbenvv opened the pull request:
# What does this PR do?

This PR adds a static cache that offloads to another device.

Fixes #32179


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@ArthurZucker @gante @n17s

## Performance tests

Performance tested it with:

- L40S (48 GB)
- Llama 3 70B (4 bit quantized)
- sdpa attention
- `torch.compile(model)`
- batch size of 1000
- sequence size of 25 (no prompt, fully generative)

and I am getting a throughput of about 535 tokens/s (OOM static).

Also with:

- batch size of 1
- sequence size of 100 (no prompt, fully generative)

which gets 10.6 tokens/s (12.8 tokens/s static).

And with:

- batch size of 1
- sequence size of 100 (fully prompt)

which does 98.8 tokens/s (106.8 tokens/s static)

And with:

- batch size of 50
- sequence size of 100 (fully prompt)

which does 939.5 tokens/s (995.6 tokens/s static)

gerbenvv committed:
Add a static cache that offloads to the CPU or other device

ArthurZucker reviewed (commented):
Already looks great! IMO would be nice to add a snippet of how to use it in the doc of the class + once #32150 is merged also add this as compatible with `torch.compile` (while the non static version won't be) 

Really like the findings you posted on the issue, will be useful for everyone I think!

ArthurZucker commented:
Also would be interesting to test this / showcase potential for huge beamsearch !

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32161). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gante commented on src/transformers/cache_utils.py:
```diff
@@ -1339,3 +1339,255 @@ def update_ssm_state(self, layer_idx: int, new_ssm_state: torch.Tensor):
+    Static cache class to be used with `torch.compile(model)` that offloads to the CPU or
+    another device.
+
+    Attributes:
```
nit: in our documentation, we place the arguments here (essentially the ones you wrote in the docstring in `__init__`)

gerbenvv replied:
Moved it over in new commit.

gante commented on src/transformers/cache_utils.py:
```diff
@@ -1339,3 +1339,255 @@ def update_ssm_state(self, layer_idx: int, new_ssm_state: torch.Tensor):
+            self._device_key_cache.append(key_cache)
+            self._device_value_cache.append(value_cache)
+
+        # For backwards compatibility.
+        self._seen_tokens = 0
```
```suggestion
```
this one is unused throughout the code

gerbenvv replied:
See comment above.

gante commented on src/transformers/cache_utils.py:
```diff
@@ -1339,3 +1339,255 @@ def update_ssm_state(self, layer_idx: int, new_ssm_state: torch.Tensor):
+
+        return k_out, v_out
+
+    def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
+        """Returns the sequence length of the cached states that were seen by the model."""
+
+        return self._seen_tokens
```
let's use the fn from the static cache (since we want to remove `self._seen_tokens`)

gerbenvv replied:
That would be a performance degradation compared to the current integer update of `self._seen_tokens`. Is there a plan to remove `get_seq_length` ? In that case, I would remove it then. Otherwise, this will be a lot slower since it will have to do (synced) CPU operation on the offloaded cache. Let me know what you think.

ArthurZucker replied:
Yeah get_seq_length is deprecated in favor of the `cache_positions` which should not need CPU operations as you only use them on device

gerbenvv replied:
Okay, but shouldn't we just remove the `get_seq_length` methods including all neccesary variables (`_seen_tokens`) once it gets removed from the API? Since there is still quite a few usages throughout the codebase.

What I meant was that `get_seq_length` will be much less performant in the meantime if I switch it over to the `StaticCache` implementation. Which if you want that in the meantime, I'm fine with as well.

ArthurZucker replied:
Ah sorry, no need to use the one from static, and actually yeah, we should probably just prevent user from using it -> no offloading for them? 
Let's go with keeping seentoken for now, add a comment saying it will be deprecated

gerbenvv replied:
Done. :+1:

gerbenvv replied:
> we should probably just prevent user from using it -> no offloading for them?
The method `get_seq_length` works fine but it's used internally still. Hence my hesitation to remove it / revert it to the slower `StaticCache` version.

gante replied:
Sounds good! 

Me and Arthur will handle the deprecation of both on all cache types + internal usage afterwards 🤗

gerbenvv replied:
@ArthurZucker Please review and merge when happy.

gante commented on src/transformers/cache_utils.py:
```diff
@@ -1339,3 +1339,255 @@ def update_ssm_state(self, layer_idx: int, new_ssm_state: torch.Tensor):
+            self.key_cache[layer_idx].zero_()
+            self.value_cache[layer_idx].zero_()
+
+    @property
+    def seen_tokens(self) -> int:
+        # For backwards compatibility.
+        return self._seen_tokens
```
```suggestion
```

gante reviewed (commented):
Very very cool! 🔥 

As Arthur wrote, this PR is now missing the complementary parts:
- An example in the docstring
- Some benchmarks in the PR, for future reference
- An integration test to prevent regressions (like the ones [here](https://github.com/huggingface/transformers/blob/main/tests/utils/test_cache_utils.py))

gerbenvv committed:
Fix PR comments, add unit-tests

gerbenvv commented:
@ArthurZucker @gante Thanks for reviewing! Made some fixes and added unit tests in a new commit. Also added the performance testing into the PR description.

gerbenvv commented:
@ArthurZucker @gante Could you help me out with the last pipeline error?

I'm running `make fix-copies` and `python utils/check_docstrings.py` locally and it works so I'm not sure what to do to fix it.

gante commented on docs/source/en/kv_cache.md:
````diff
@@ -238,6 +238,24 @@ For more examples with Static Cache and JIT compilation, take a look at [StaticC
 "Hello, my name is [Your Name], and I am a [Your Profession] with [Number of Years] of"
 ```
 
+Like [`~OffloadedCache`] exists for offloading a "DynamicCache", there is also an offloaded static cache. Just
````
```suggestion
### Offloaded Static Cache

Like [`~OffloadedCache`] exists for offloading a "DynamicCache", there is also an offloaded static cache. Just
```

I think it deserves a subsection of its own 🤗

gerbenvv replied:
Will do, also, I will add it to the overview table.

gerbenvv replied:
Done! Although I do wonder what `Initialization Recommended` means in the overview table?

ArthurZucker replied:
Whether it should be init outside `generate` and passed to generate or not cc @zucchini-nlp on this!

zucchini-nlp replied:
Some cache classes are recommended to initialize outside generation, e.g. StaticCache with compilation had some issues when we initialized cache while compiling.

Also, some cache types are not handled automatically by our API, e.g. SinkCache so the user has no option as to initialize and pass `past_key_values`

gante reviewed (approved):
LGTM 🤗 

Thank you for the cool feature and for iterating with us!

ArthurZucker reviewed (approved):
Well sorry for the late review, very very nice, let's make sure the slow tests pass and good to go! Can you try to run them locally? 🤗

ArthurZucker commented:
@gerbenvv I can also merge like this and commit later if you are busy!

gerbenvv commented:
> @gerbenvv I can also merge like this and commit later if you are busy!

Yeah trying to run the tests now `RUN_SLOW=1 pytest tests/utils/test_cache_utils.py` but it's taking quite a while.
I also tried to install the full dev environment but it failed to install `av` due to some compile error.

I'll try running just the file `pytest tests/utils/test_cache_utils.py` and if that doesn't work for me, the tests that I have changed/added. Since it is basically a new class, that should be fine.

gerbenvv commented:
Hmm, the tests have literally crashed the whole machine. Is it supposed to use all the GPUs on the machine?

I am struggling a bit to run this.

gerbenvv commented:
`RUN_SLOW=1 pytest tests/utils/test_cache_utils.py` gave me:

```
========================================================================================================= short test summary info ==========================================================================================================
FAILED tests/utils/test_cache_utils.py::CacheTest::test_static_cache_exportability - OSError: You are trying to access a gated repo.                                                                                                        
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_dynamic_cache_batched - OSError: You are trying to access a gated repo.                                                                                                  
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_dynamic_cache_beam_search - OSError: You are trying to access a gated repo.                                                                                              
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_dynamic_cache_hard - OSError: You are trying to access a gated repo.                                                                                                     
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_hybrid_cache_n_sequences - OSError: You are trying to access a gated repo.                                                                                               
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_offloaded_cache_equivalent_to_dynamic_cache - ImportError: Using `low_cpu_mem_usage=True` or a `device_map` requires Accelerate: `pip install accelerate`                
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_offloaded_cache_uses_less_memory_than_dynamic_cache - ImportError: Using `low_cpu_mem_usage=True` or a `device_map` requires Accelerate: `pip install accelerate`        
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_sink_cache_iterative_prompts - ImportError: Using `low_cpu_mem_usage=True` or a `device_map` requires Accelerate: `pip install accelerate`                               
===================================================================================== 8 failed, 14 passed, 2 skipped, 24 warnings in 627.82s (0:10:27) ====================================================================================
```

Then I tried to fix those errors by passing the token & installing `accelerate` but then the machine crashed. But we at least know that:

```
test_static_cache_greedy_decoding_pad_left
test_static_cache_greedy_decoding_pad_right
test_static_cache_extra_left_padding
```

ran succesful and those were the ones that I have changed.

gerbenvv commented:
Okay, making progress ;-)

New output of `CUDA_VISIBLE_DEVICES=0 RUN_SLOW=1 pytest tests/utils/test_cache_utils.py` is:

```
========================================================================================================= short test summary info ==========================================================================================================
FAILED tests/utils/test_cache_utils.py::CacheTest::test_static_cache_exportability - torch._export.verifier.SpecViolationError: Node.meta _enter_autocast is missing val field.
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_hybrid_cache_n_sequences - OSError: You are trying to access a gated repo.
FAILED tests/utils/test_cache_utils.py::CacheIntegrationTest::test_sink_cache_iterative_prompts - AssertionError: False is not true
===================================================================================== 3 failed, 19 passed, 2 skipped, 28 warnings in 756.35s (0:12:36) =====================================================================================
```

So no tests regarding the static, dynamic or offloading caches are failing.

ArthurZucker commented:
Yep let's go! 🔥
